lib/concurrent/configuration.rb

```diff
@@ -1,124 +1,296 @@
+  end
+
+  # Defines if *ALL* executors should be auto-terminated with an
+  # `at_exit` callback. When set to `false` it will be the application
+  # programmer's responsibility to ensure that *all* thread pools,
+  # including the global thread pools, are shutdown properly prior to
+  # application exit.
+  #
+  # @note Only change this option if you know what you are doing!
+  #   When this is set to true (the default) then `at_exit` handlers
+  #   will be registered automatically for *all* thread pools to
+  #   ensure that they are shutdown when the application ends. When
+  #   changed to false, the `at_exit` handlers will be circumvented
+  #   for *all* Concurrent Ruby thread pools running within the
+  #   application. Even those created within other gems used by the
+  #   application. This method should *never* be called from within a
+  #   gem. It should *only* be used from within the main application.
+  #   And even then it should be used only when necessary.
+  #
+  # @return [Boolean] true when *all* thread pools will auto-terminate on
+  #   application exit using an `at_exit` handler; false when no auto-termination
+  #   will occur.
+  def self.auto_terminate_all_executors?
+    @@auto_terminate_all_executors.value
+  end
+
+  # Global thread pool optimized for short, fast *operations*.
+  #
+  # @return [ThreadPoolExecutor] the thread pool
+  def self.global_fast_executor
+    @@global_fast_executor.value
+  end
+
+  # Global thread pool optimized for long, blocking (IO) *tasks*.
+  #
+  # @return [ThreadPoolExecutor] the thread pool
+  def self.global_io_executor
+    @@global_io_executor.value
+  end
+
+  # Global thread pool user for global *timers*.
+  #
+  # @return [Concurrent::TimerSet] the thread pool
+  #
+  # @see Concurrent::timer
+  def self.global_timer_set
+    @@global_timer_set.value
+  end
+
+  def self.shutdown_global_executors
+    global_fast_executor.shutdown
+    global_io_executor.shutdown
+    global_timer_set.shutdown
+  end
+
+  def self.kill_global_executors
+    global_fast_executor.kill
+    global_io_executor.kill
+    global_timer_set.kill
+  end
+
+  def self.wait_for_global_executors_termination(timeout = nil)
+    latch = CountDownLatch.new(3)
+    [ global_fast_executor, global_io_executor, global_timer_set ].each do |executor|
+      Thread.new{ executor.wait_for_termination(timeout); latch.count_down }
+    end
+    latch.wait(timeout)
+  end
+
+  def self.new_fast_executor(opts = {})
+    FixedThreadPool.new(
+      [2, Concurrent.processor_count].max,
+      stop_on_exit:    opts.fetch(:stop_on_exit, true),
+      idletime:        60,          # 1 minute
+      max_queue:       0,           # unlimited
+      fallback_policy: :caller_runs # shouldn't matter -- 0 max queue
+    )
+  end
+
+  def self.new_io_executor(opts = {})
+    ThreadPoolExecutor.new(
+      min_threads: [2, Concurrent.processor_count].max,
+      max_threads: Concurrent.processor_count * 100,
```

Comment:
we should consider same default as on java pool http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/util/concurrent/Executors.java#Executors.newCachedThreadPool%28%29, unlimited threads. To prevent deadlock when all available tasks are blocked by work waiting in the queue, etc. I think if there is too many work being posted the the pool it needs to be contained differently.